lib/Optimizer/Transforms/UnitarySynthesis.cpp

```diff
@@ -0,0 +1,212 @@
+
+    /// If the replacement function doesn't exist, create it here
+    if (!parentModule.lookupSymbol<LLVM::LLVMFuncOp>(funcName)) {
+
+      auto unitary = cudaq::readGlobalConstantArray(globalOp);
+
+      /// TODO: Expand the logic to decompose upto 4-qubit operations
+      if (unitary.size() != 4)
+        return customOp.emitError(
+            "Decomposition of only single qubit custom operations supported.");
+
+      /// Controls are handled via apply specialization
+      auto funcTy =
+          FunctionType::get(parentModule.getContext(), targets.getTypes(), {});
+      auto insPt = rewriter.saveInsertionPoint();
+      rewriter.setInsertionPointToStart(parentModule.getBody());
+      auto func = rewriter.create<func::FuncOp>(parentModule->getLoc(),
+                                                funcName, funcTy);
+      func.setPrivate();
+      rewriter.restoreInsertionPoint(insPt);
+
+      insPt = rewriter.saveInsertionPoint();
+      auto *block = func.addEntryBlock();
+      rewriter.setInsertionPointToStart(block);
+
+      /// Use Euler angle decomposition for single qubit operation
+      auto zyz = BasisZYZ(unitary);
+      zyz.decompose();
+
+      /// For 1-qubit operation, apply on 'all' the targets
+      auto arguments = func.getArguments();
+      auto floatTy = cast<FloatType>(rewriter.getF64Type());
+
+      /// Ignore angles less than some threshold
+      double epsilon = 1e-9;
+      auto isAboveThreshold = [&](auto value) {
+        return std::abs(value) > epsilon;
+      };
+
+      if (isAboveThreshold(zyz.angles.alpha)) {
+        auto alpha = cudaq::opt::factory::createFloatConstant(
+            loc, rewriter, zyz.angles.alpha, floatTy);
+        rewriter.create<quake::RzOp>(loc, alpha, ValueRange{}, arguments);
+      }
+
+      if (isAboveThreshold(zyz.angles.beta)) {
+        auto beta = cudaq::opt::factory::createFloatConstant(
+            loc, rewriter, zyz.angles.beta, floatTy);
+        rewriter.create<quake::RyOp>(loc, beta, ValueRange{}, arguments);
+      }
+
+      if (isAboveThreshold(zyz.angles.gamma)) {
+        auto gamma = cudaq::opt::factory::createFloatConstant(
+            loc, rewriter, zyz.angles.gamma, floatTy);
+        rewriter.create<quake::RzOp>(loc, gamma, ValueRange{}, arguments);
+      }
+
+      /// NOTE: Typically global phase can be ignored but, if this decomposition
+      /// is applied in a kernel that is called with `cudaq::control`, the
+      /// global phase will become a local phase and give a wrong result if we
+      /// don't keep track of that.
+      if (isAboveThreshold(zyz.phase)) {
+        auto phase = cudaq::opt::factory::createFloatConstant(
+            loc, rewriter, zyz.phase, floatTy);
+        Value negPhase = rewriter.create<arith::NegFOp>(loc, phase);
+        rewriter.create<quake::R1Op>(loc, phase, ValueRange{}, arguments[0]);
+        rewriter.create<quake::RzOp>(loc, negPhase, ValueRange{}, arguments[0]);
```

Comment:
Do we have a test for this?
My back-of-the-envelope calculation is that a R1-Rz pair results in a theta/2 global phase, i.e., we need to multiply the angle by 2.

Reply:
In my current tests, applying global phase had no effect on the results. I will add a test such that this is tested.

Reply:
Covered by newly added tests in ecb3c0c1119c0cb418106625b03a2ca36342012b